src/midi/midi_fluidsynth.cpp

```diff
@@ -163,21 +205,39 @@ void MidiHandlerFluidsynth::PlaySysex(uint8_t *sysex, size_t len)
 	fluid_synth_sysex(synth.get(), data, n, nullptr, nullptr, nullptr, false);
 }
 
-void MidiHandlerFluidsynth::mixer_callback(uint16_t frames)
+void MidiHandlerFluidsynth::PrintStats()
+{
+	// Normally prescale is simply a float-multiplier such as 0.5, 1.0, etc.
+	// However in the case of FluidSynth, it produces 32-bit floats between
+	// -1.0 and +1.0, therefore we scale those up to the 16-bit integer range
+	// in addition to the mixer's FSYNTH levels. Before printing statistics,
+	// we need to back-out this integer multiplier.
+	prescale_level.left /= INT16_MAX;
+	prescale_level.right /= INT16_MAX;
+	soft_limiter.PrintStats();
+}
+
+void MidiHandlerFluidsynth::MixerCallBack(uint16_t frames)
 {
-	constexpr uint16_t expected_max_frames = (96000 / 1000) + 4;
-	int16_t data[expected_max_frames * 2]; // two channels per frame
+	constexpr uint16_t max_samples = expected_max_frames * 2; // two channels per frame
+	std::array<float, max_samples> accumulator = {{0}};
+	std::array<int16_t, max_samples> scaled = {{0}};
```

Comment:
I think we might prefer to keep these buffers uninitialized? We *know* fluidsynth and soft_limiter will fill them?

I don't understand why float array is called `accumulator`…

I don't like that `expected_max_frames` leaked into surrounding context… previously it was scoped to this singe buffer allocation (with +4 being specific to buffer size and type - as to land on SIMD-friendly buffer boundary).

*note to myself*: double-check if loop is still expected to go through 1 iteration in normal case.

Reply:
Right; I don't think there's any need to pre-zero the arrays. Will drop that. 

Accumulator: this is where FluidSynth accumulates the final float amplitudes from all the active parallel voices. That said, we don't need to be that specific and it would be more universally understood if I just called it 'output'. (Will update this). 

Yes.. I hit link errors with the max output frames when assigned as a static constexpr in the header.  I will revert that and pass you the error to see if you can solve it.  I fully agree this is an ugly and wrong approach.

Reply:
@kcgen Do you think we should merge the code in as it is and improve it in future PRs?

After looking at it again I have no further comments. And I think we should move discussion about existing problems to the main thread to draw in more testers.